### ISSUE TITLE
Dispose pattern not implemented properly

### ISSUE THREAD

**Issue by darkurse** [lukevp/ESC-POS-.NET#16]
Hi there,

first of, thank you very much for your project !!

Regarding the Dispose pattern. NetworkPrinter doesn't implement a safe disposable pattern.

When using the simple following code, things go kaboum

```C#
using (var p = new NetworkPrinter())
{
   // [....]
}
```

That's because you also call Dispose() in the finalizer without checking if resource has already been disposed.

EDIT: I had to use a `GC.SuppressFinalize(p);` to prevent GC disposing twice the resource

---
EDIT2 : An even worst offender

In the case an printer address is invalid or unreachable

```C#
public class NetworkPrinter : BasePrinter, IDisposable
{
        // ...

        private void Connect()
        {
            _socket = new Socket(_endPoint.AddressFamily, SocketType.Stream, ProtocolType.Tcp);

           // What if the line below cannot connect properly ?
            _socket.Connect(_endPoint);

           // Then what's below never gets initialized and is still set to null
            _sockStream = new NetworkStream(_socket);
            _writer = new BinaryWriter(_sockStream);
            _reader = new BinaryReader(_sockStream);
        }

        public void Dispose()
        {
            _writer.Close(); // <-- Crash with NullReferenceException
            _reader.Close(); // <-- Crash with NullReferenceException
            _socket.Close();
            _socket.Dispose();
        }
}
```

**Comment by lukevp** [comment]
Good points and I'm glad you are finding the library useful!
To be honest I'm not sure why they don't implement dispose properly. Is this something youd be willing to test and make a PR for?  We should fix this in all of the printers since they all have that same issue. 
How about using the null conditional operator for the close calls?

**Comment by darkurse** [comment]
Hi Luke, I had to deliver this fix quickly so I actually forked your project to do so and put the nuget package on private feed.

However I can push you a PR later on. FYI, that's my fix

```C#
public void Dispose()
{
    if (_socket != null)
    {
        _writer?.Close();
        _reader?.Close();
        _socket.Close();
        _socket = null;

        GC.SuppressFinalize(this);
    }
}
```

1. Even if Dispose is called twice manually, it won't attempt to release the socket twice.
2. Writer and Reader are Null Reference protected
3. If dispose was called manually, there is no longer need to call the finalizer.
